src/modules/schnorr_adaptor/main_impl.h

```diff
@@ -0,0 +1,286 @@
+             116, 128,  68,  31, 144, 186,  37, 196,
+             136, 244,  97, 199,  11,  94, 165, 220,
+             170, 247, 175, 105, 39,  10, 165,  20
+        };
+        for (i = 0; i < 32; i++) {
+            masked_key[i] = key32[i] ^ ZERO_MASK[i];
+        }
+    }
+
+    /* Tag the hash with algo which is important to avoid nonce reuse across
+     * algorithms. If this nonce function is used in BIP-340 signing as defined
+     * in the spec, an optimized tagging implementation is used. */
+    if (algolen == sizeof(bip340_algo)
+            && secp256k1_memcmp_var(algo, bip340_algo, algolen) == 0) {
+        secp256k1_nonce_function_bip340_sha256_tagged(&sha);
+    } else {
+        secp256k1_sha256_initialize_tagged(&sha, algo, algolen);
+    }
+
+    /* Hash masked-key||pk||msg using the tagged hash as per the spec */
+    secp256k1_sha256_write(&sha, masked_key, 32);
+    secp256k1_sha256_write(&sha, xonly_t32, 32);
+    secp256k1_sha256_write(&sha, xonly_pk32, 32);
+    secp256k1_sha256_write(&sha, msg32, 32);
+    secp256k1_sha256_finalize(&sha, nonce32);
+    return 1;
+}
+
+const secp256k1_adaptor_nonce_function_hardened secp256k1_adaptor_nonce_function_bip340 = adaptor_nonce_function_bip340;
+
+static int secp256k1_schnorr_adaptor_presign_internal(const secp256k1_context *ctx, unsigned char *sig65, const unsigned char *msg32, const secp256k1_keypair *keypair, secp256k1_adaptor_nonce_function_hardened noncefp, const unsigned char *t33, void *ndata) {
+    secp256k1_scalar sk;
+    secp256k1_scalar e;
+    secp256k1_scalar k;
+    secp256k1_gej rj;
+    secp256k1_gej r0j;
+    secp256k1_ge pk;
+    secp256k1_ge r;
+    secp256k1_ge r0;
+    secp256k1_ge t;
+    unsigned char nonce32[32] = {0};
+    unsigned char pk_buf[32];
+    unsigned char seckey[32];
+    size_t size = 33;
+    size_t msglen = 32;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(sig65 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(keypair != NULL);
+    ARG_CHECK(t33 != NULL);
+
+    if (noncefp == NULL) {
+        noncefp = secp256k1_adaptor_nonce_function_bip340;
+    }
+
+    ret &= secp256k1_keypair_load(ctx, &sk, &pk, keypair);
+
+    if (secp256k1_fe_is_odd(&pk.y)) {
+        secp256k1_scalar_negate(&sk, &sk);
+    }
+
+    /* d */
+    secp256k1_scalar_get_b32(seckey, &sk);
+    /* bytes_from_point(P) */
+    secp256k1_fe_get_b32(pk_buf, &pk.x);
+
+    ret &= !!noncefp(nonce32, msg32, seckey, &t33[1], pk_buf, bip340_algo, sizeof(bip340_algo), ndata);
+    /* k0 */
+    secp256k1_scalar_set_b32(&k, nonce32, NULL);
+    ret &= !secp256k1_scalar_is_zero(&k);
+    secp256k1_scalar_cmov(&k, &secp256k1_scalar_one, !ret);
+
+    /* R = k0*G */
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &rj, &k);
+    secp256k1_ge_set_gej(&r, &rj);
+
+    /* T = cpoint(T) */
+    ret &= !!secp256k1_eckey_pubkey_parse(&t, t33, 33);
```

Comment:
I thought this is a convention for using !! instead of nothing

Reply:
The other parts of libsecp use `!!` only for `noncefp` (and not for `parse` and `serialize`). So, we can remove this here imo.

Reply:
`!!` is an idiom for ensuring that the value is either `0` or `1` (and not some other non-zero value). Recall that C doesn't have a proper bool type. 

We use `!!` for valued returned by `noncefp` because `noncefp` is provided by the user, so we don't control it. For the functions in our code base, our usual convention is that the called function is responsible for returning only `0` and `1`. So yes, strictly speaking, `!!` is not necessary here, but it also won't hurt. I'd keep it.